### ISSUE TITLE
New rule request: Enforce a maximum number of lines to try

### ISSUE THREAD

**Issue by yvele** [sindresorhus/eslint-plugin-unicorn#651]
The rule name may be called `unicorn/try-maximum-lines`.
The idea is to avoid `try` on whole code blocks and encourage the developer to explicitly snipe the code that is expecting to throw.

Should we add a parameter to customize the maximum number of line or only set to to 1 line of code? 🤔 

Note that I'm not a native english speaker so please fix the rule/option names.

# Enforce a maximum number of lines to try

Applies to `try` clauses.

The maximum number of triable lines of code configurable, but defaults to 1.

This rule is not fixable.

## Fail

```js
try {
  foo();
  bar();
} catch (err) {
  // …
}
```

## Pass

```js
try {
  foobar();
} catch (err) {
  // …
}

## Options

### `case`

Type: `number`

You can set the maximum number of lines like this:

```js
"unicorn/try-maximum-line": [
	"error",
	{
		"maximumNumberOfLine": 2
	}
]
```

**Comment by yvele** [comment]
Should this pass? 🤔 

```js
try {
  return doSomething() || doSomethingElse();
} catch (err) {
  // …
}
```

This may pass. I think we can limit this rule to count the number of lines, not the number of instructions.

**Comment by fregante** [comment]
I don't like this but I think it can help. I usually stick a lot of code inside `try` just because I don't want to do use `let`:

```js
let a;
try {
	a = maybethrows();
} catch {
	// handle
}

twistIt(a);
bopIt(a);
pullIt(a);
```

and instead I use `const`:

```js
let a;
try {
	const a = maybethrows();
	twistIt(a);
	bopIt(a);
	pullIt(a);
} catch {
	// handle
}
```

**Comment by yvele** [comment]
The purpose of that rule is to avoid having bunch of code inside a `try` 🤷‍♂ 

I don't see the problem with `let` 🤔

**Comment by fregante** [comment]
> I don't see the problem with `let` 🤔

`const` is preferred because it forces you to double-check any changes you make to the variable.
Also `const` is one line while the alternative is 2:

```js
let a;
try {
	a = maybethrows();
```

**Comment by sindresorhus** [comment]
I'm lukewarm on this. I fear it will be more annoying than useful.

For example, here, I don't want to expose an intermediate variable outside the try scope.

```js
let foo
try {
  const bar = getBar();
  foo = foo.slice(1, 2);
} catch (err) {
  // …
}

console.log(foo);
```

Maybe it should instead enforce *only one function call inside "try"*?

**Comment by fregante** [comment]
> Maybe it should instead enforce _only one function call inside "try"_?

`.slice` is also a function

**Comment by sindresorhus** [comment]
That's the problem though. What is a sensible line limit? There's no universal answer. I used to have the built-in ESLint line limit rules enabled, but ended up having to turn them off as measuring things by line numbers is too limiting and annoying.

**Comment by sindresorhus** [comment]
The example was obviously artificial just to show that sometimes it makes sense to use multiple lines.

Here are some examples I grepped from my projects:

```js
try {
	fs.statSync('/.dockerenv');
	return true;
} catch (_) {
	return false;
}
```

```js
test('non-integer count should throw exception', async t => {
	try {
		await beeper(1.5);
		t.fail();
	} catch (error) {
		t.pass();
	}
});
```

```js
try {
	const value = await Promise.all([total, element.value]);
	next(reducer(value[0], value[1], index++));
} catch (error) {
	reject(error);
}
```

```js
try {
	const stats = await promisify(fs[fsStatType])(filePath);
	return stats[statsMethodName]();
} catch (error) {
	if (error.code === 'ENOENT') {
		return false;
	}

	throw error;
}
```

```js
try {
	await subprocess;
	callback();
} catch (_) {
	callback(new PluginError('gulp-ava', 'One or more tests failed'));
}
```

```js
try {
	for (const handler of this._cancelHandlers) {
		handler();
	}
} catch (error) {
	this._reject(error);
}
```

```js
try {
	const {stdout} = await execa(shell || defaultShell, args);
	return parseEnv(stdout);
} catch (error) {
	if (shell) {
		throw error;
	} else {
		return process.env;
	}
}
```

```js
const handleKill = async input => {
	try {
		input = await parseInput(input);

		if (input === process.pid) {
			return;
		}

		if (input === 'node') {
			const processes = await psList();
			await Promise.all(processes.map(async ps => {
				if (ps.name === 'node' && ps.pid !== process.pid) {
					await kill(ps.pid, options);
				}
			}));
			return;
		}

		await kill(input, options);
	} catch (error) {
		if (!exists.get(input)) {
			errors.push(`Killing process ${input} failed: Process doesn't exist`);
			return;
		}

		errors.push(`Killing process ${input} failed: ${error.message.replace(/.*\n/, '').replace(/kill: \d+: /, '').trim()}`);
	}
};
```

**Comment by sindresorhus** [comment]
How would it handle nested try/catch?

**Comment by yvele** [comment]
> The example was obviously artificial just to show that sometimes it makes sense to use multiple lines.
> 
> Here are some examples I grepped from my projects:


<br>

---------

<br>

> ```js
> try {
> 	fs.statSync('/.dockerenv');
> 	return true;
> } catch (_) {
> 	return false;
> }
> ```


```js
try {
	fs.statSync('/.dockerenv');
} catch (_) {
	return false;
}
return true;
```

<br>

---------

<br>

> ```js
> test('non-integer count should throw exception', async t => {
> 	try {
> 		await beeper(1.5);
> 		t.fail();
> 	} catch (error) {
> 		t.pass();
> 	}
> });
> ```

If `t.fail()` it self throws, you don't want to catch it and make the test to pass.

You may use [`expect().toThrow()`](https://jestjs.io/docs/en/expect#tothrowerror) test assertion (Jest as an example). I'm pretty sure you testing framework as a proper solution to assert errors are thrown.

```js
test('non-integer count should throw exception', async () => {
    await expect(
	  async () => beeper(1.5)
    ).rejects.toThrow();
});
```

<br>

---------

<br>

> ```js
> try {
> 	const value = await Promise.all([total, element.value]);
> 	next(reducer(value[0], value[1], index++));
> } catch (error) {
> 	reject(error);
> }
> ```


```js
let value;
try {
	value = await Promise.all([total, element.value]);
} catch (error) {
	reject(error);
    return;
}
next(reducer(value[0], value[1], index++));
```

<br>

---------

<br>

> ```js
> try {
> 	const stats = await promisify(fs[fsStatType])(filePath);
> 	return stats[statsMethodName]();
> } catch (error) {
> 	if (error.code === 'ENOENT') {
> 		return false;
> 	}
> 
> 	throw error;
> }
> ```

Not sure what exact instruction you actually are trying to `try` (that's the whole point of the rule). But let's say:

```js

// I assume we don't want to catch this
const statFunc = promisify(fs[fsStatType]);

let stats;
try {
	stats = await statFunc(filePath);
} catch (error) {
	if (error.code === 'ENOENT') {
		return false;
	}

	throw error;
}
return stats[statsMethodName]();
```

<br>

---------

<br>

> ```js
> try {
> 	await subprocess;
> 	callback();
> } catch (_) {
> 	callback(new PluginError('gulp-ava', 'One or more tests failed'));
> }
> ```


```js
try {
	await subprocess;
} catch (_) {
	callback(new PluginError('gulp-ava', 'One or more tests failed'));
    return;
}
callback();
```

<br>

---------

<br>

> ```js
> try {
> 	for (const handler of this._cancelHandlers) {
> 		handler();
> 	}
> } catch (error) {
> 	this._reject(error);
> }
> ```


```js
for (const handler of this._cancelHandlers) {
   try {
    	handler();
   } catch (error) {
		this._reject(error);
		return;
  }
}
```

Or you may wrap your loop in a dedicated function


<br>

---------

<br>

> ```js
> try {
> 	const {stdout} = await execa(shell || defaultShell, args);
> 	return parseEnv(stdout);
> } catch (error) {
> 	if (shell) {
> 		throw error;
> 	} else {
> 		return process.env;
> 	}
> }
> ```


I assume you don't want to catch `parseEnv` 🤷‍♂ (another advantage of this rule is to spot exactly what is expected to throw).

```js
let stdout;
try {
	stdout = (await execa(shell || defaultShell, args)).stdout;
} catch (error) {
	if (shell) {
		throw error;
	} else {
		return process.env;
	}
}
return parseEnv(stdout);
```

<br>

---------

<br>

> ```js
> const handleKill = async input => {
> 	try {
> 		input = await parseInput(input);
> 
> 		if (input === process.pid) {
> 			return;
> 		}
> 
> 		if (input === 'node') {
> 			const processes = await psList();
> 			await Promise.all(processes.map(async ps => {
> 				if (ps.name === 'node' && ps.pid !== process.pid) {
> 					await kill(ps.pid, options);
> 				}
> 			}));
> 			return;
> 		}
> 
> 		await kill(input, options);
> 	} catch (error) {
> 		if (!exists.get(input)) {
> 			errors.push(`Killing process ${input} failed: Process doesn't exist`);
> 			return;
> 		}
> 
> 		errors.push(`Killing process ${input} failed: ${error.message.replace(/.*\n/, '').replace(/kill: \d+: /, '').trim()}`);
> 	}
> };
> ```

I'm pretty sure this may be refactored with function that wraps to cod that is in `try`

**Comment by yvele** [comment]
> How would it handle nested try/catch?

The same way.. only n lines of code allowed in `try`

**Comment by sindresorhus** [comment]
```js
try {
	fs.statSync('/.dockerenv');
} catch (_) {
	return false;
}
return true;
```

While this works, I still prefer early return. As I do with `if` statements. What if there was more logic after the try-catch. That would make it more difficult to handle.

---

```js
for (const handler of this._cancelHandlers) {
   try {
    	handler();
   } catch (error) {
		this._reject(error);
		return;
  }
}
```

Try-catch comes with certain overhead, so it would be slower to put it in the loop.

**Comment by sindresorhus** [comment]
@yvele Instead of lines, how about we limit it to one statement? And maybe also an option to allow a `return` statement at the end in addition.

I still feel like "number of lines" is too naive.

For example, the below is still one statement:

```js
try {
	result = foo
		.someMethodChainMethod1()
		.someMethodChainMethod2()
		.someMethodChainMethod3()
} catch {}
```

**Comment by yvele** [comment]
> @yvele Instead of lines, how about we limit it to one statement?
>
> ...
>
> I still feel like "number of lines" is too naive.
>
> For example, the below is still one statement:
> 
> ```js
> try {
> 	result = foo
> 		.someMethodChainMethod1()
> 		.someMethodChainMethod2()
> 		.someMethodChainMethod3()
> } catch {}
> ```

Yep I [initially suggested a limit to one **instruction**](https://github.com/sindresorhus/eslint-plugin-unicorn/issues/651#issuecomment-606549289), but one **statement** is ok



> And maybe also an option to allow a `return` statement at the end in addition.

Why not. I personally will not use it.


> Try-catch comes with certain overhead, so it would be slower to put it in the loop.

I know that catching error comes with lots of overhead (building the call stack, etc.) but I wasn't aware that simply using `try` also comes with overhead even when no error is thrown. 🤔 

Do you have any documentation to share about that `try` overhead? I tried Googling https://www.google.com/search?q=javascript+try+overhead but nothing relevant.

**Comment by sindresorhus** [comment]
> Do you have any documentation to share about that try overhead? I tried Googling google.com/search?q=javascript+try+overhead but nothing relevant.

🤷‍♂️ Just from memory. Might not even be valid anymore.

**Comment by sindresorhus** [comment]
This is now accepted. PR welcome.

**Comment by sindresorhus** [comment]
Name: `try-maximum-statements`

**Comment by fisker** [comment]
I like this, should be easy to implement.

**Comment by fisker** [comment]
@yvele You suggest 


```js
let foo
try {
  const bar = getBar();
  foo = bar.slice(1, 2);
} catch (err) {
  // …
}

console.log(foo);
```

Into:

```js
let foo
try {
 foo = getBar().slice(1, 2);
} catch (err) {
  // …
}

console.log(foo);
```

But if it's

```js
let foo
try {
  const bar = getBar();
  foo = bar.slice(0, Math.floor(bar.length / 2));
} catch (err) {
  // …
}

console.log(foo);
```

**Comment by yvele** [comment]
The fixes I've suggested are hypothetical.

If there is multiple statements in the `try` we cannot determine which exact statement is expected to throw. It's also possible (even unlikely) that multiple statements are expected to throw (in which case statements must be split into dedicated `try`).

This rule cannot be fixable.

Does that make sense?

**Comment by futpib** [comment]
Another possibility is that developers won't target the right statement (if there is one) and instead group statements into a single one:

```js
const doTwoThings = () => {
  doSomething();
  doSomethingElse();
};

try {
  doTwoThings();
}
```

I kinda like this fact except no rule can ensure that the new function has a meaningful name (and that it can have one, given that it was created to dodge a syntactic rule).

This also should that the number of statements does not mean much really.

<hr/>

What about this code:

```js
try {
	const result = await getResult();
	return result; // Could be `subscriber.next(result);`, etc.
}
```

Do I have to create an outer mutable variable and assign to it in `try`?

```js
let result;

try {
	result = await getResult();
} catch (...) {...}

return result;
```

Should check if `result` is not `undefined` before returning? What if `getResult` returned `undefined`? Should I add another flag variable just to get the same result as above?

I much prefer the first option.

<hr/>

I hate when there is redundant code in a `try` block, but I don't think this rule will help with it enough to bear it's annoyance.

Maybe this rule is useful in some projects (I imagine), but not generally (at least coming from my experience).

**Comment by yvele** [comment]
> What about this code:
> 
> ```js
> try {
> 	const result = await getResult();
> 	return result; // Could be `subscriber.next(result);`, etc.
> }
> ```

```js
try {
  return await getResult();
}
```

🤷‍♂️

**Comment by papb** [comment]
I also like this rule, however I think allowing an extra `return` statement that can be statically guaranteed to not throw (such as `return true`) is a must.

**Comment by yvele** [comment]
> I also like this rule, however I think allowing an extra `return` statement that can be statically guaranteed to not throw (such as `return true`) is a must.

Yes maybe we can add an option not to count `return` as a statement 🤔 whether it may throw or not.

**Comment by fisker** [comment]
Let's use the same algorithm from [`complexity`](https://github.com/eslint/eslint/blob/master/lib/rules/complexity.js) rule?

**Comment by sindresorhus** [comment]
@fisker Sounds good. Let's start with that and we can potentially iterate based on real world experience with it.

**Comment by papb** [comment]
@fisker @sindresorhus What are your opinions on allowing an extra `return someValue;` apart from the single statement?

**Comment by fisker** [comment]
Hard to tell before check real world cases.
